### PR TITLE
feat(ingestion): adds platform instance capability to glue connector

### PR DESCRIPTION
**Pull request by sgomezvillamor**
This is enabling platform instance capability to the glue connector.

Additionally, `underlying_platform` field has been deprecated in favour of the standard `platform` one.

During the development of this task, it was noted that other connectors were badly setting the factory as the instance for the container key. This was fixed too and so the scope of the PR went a little bit beyond the glue connector.

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by sgomezvillamor** [comment]
Just for the curious, the new test replicates the previous one with the simple addition of the `platform_instance` config. This is noted in the diff of the golden files:

```
(venv) ➜  sgomezvillamor-datahub git:(glue-platform-instance) ✗ diff metadata-ingestion/tests/unit/glue/glue_mces_golden.json metadata-ingestion/tests/unit/glue/glue_mces_platform_instance_golden.json 
6c6
<         "urn": "urn:li:dataset:(urn:li:dataPlatform:glue,flights-database.avro,PROD)",
---
>         "urn": "urn:li:dataset:(urn:li:dataPlatform:glue,some_instance_name.flights-database.avro,PROD)",
236c236
<         "urn": "urn:li:dataset:(urn:li:dataPlatform:glue,test-database.test_jsons_markers,PROD)",
---
>         "urn": "urn:li:dataset:(urn:li:dataPlatform:glue,some_instance_name.test-database.test_jsons_markers,PROD)",
403c403
<         "urn": "urn:li:dataset:(urn:li:dataPlatform:glue,test-database.test_parquet,PROD)",
---
>         "urn": "urn:li:dataset:(urn:li:dataPlatform:glue,some_instance_name.test-database.test_parquet,PROD)",
734c734
<                 "urn:li:dataset:(urn:li:dataPlatform:glue,flights-database.avro,PROD)"
---
>                 "urn:li:dataset:(urn:li:dataPlatform:glue,some_instance_name.flights-database.avro,PROD)"
959c959
<                 "urn:li:dataset:(urn:li:dataPlatform:glue,test-database.test_parquet,PROD)"
---
>                 "urn:li:dataset:(urn:li:dataPlatform:glue,some_instance_name.test-database.test_parquet,PROD)"
```

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:50]: nit: 
```python 
VALID_PLATFORMS = [DEFAULT_PLATFORM, "athena"]
```

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:97]: Delete this validator. This connector should be able to rely on any validation added for `platform_instance` in the future by `PlatformSourceConfigBase`.

File: metadata-ingestion/tests/unit/test_glue_source.py
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_glue_source.py:70]: This is interesting! What values will be supplied to the first two params, `tmp_path` & `pytestconfig`?

File: metadata-ingestion/tests/unit/test_glue_source.py
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_glue_source.py:81]: Adding type annotations would be nice!

File: metadata-ingestion/tests/unit/test_glue_source.py
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_glue_source.py:178]: Pretty nice coverage for config! :-)

File: metadata-ingestion/tests/unit/test_glue_source_stubs.py
Comment (rslanka) [inline metadata-ingestion/tests/unit/test_glue_source_stubs.py:762]: Adding type annotations would be nice!

**Comment by rslanka** [comment]
Looks great overall! Please address the minor comments if possible.

File: metadata-ingestion/tests/unit/test_glue_source.py
Comment (sgomezvillamor) [reply metadata-ingestion/tests/unit/test_glue_source.py:70]: To be honest, I haven't looked into them, just assumed they were provided by `pytest` 😅

Looking into this now, I have found that there are some built-in fixtures supporting them: 
`tmp_path` https://docs.pytest.org/en/6.2.x/tmpdir.html
`pytestconfig` https://docs.pytest.org/en/latest/reference/reference.html#pytestconfig

**Comment by sgomezvillamor** [comment]
Thanks for the review @rslanka, addressed all your comments in 0ecf83d
Merging master and fixing conflicts...

**Comment by rslanka** [comment]
This is good to go once the merge with master is pushed.

**Comment by sgomezvillamor** [comment]
@rslanka I have merged master and it should pass tests now. However, I noted that domains and containers were added and I have some doubts. If you have a look at the `glue_mces_platform_instance_golden.json`, the generated mcp for containers look as follows:

```
  {
    "auditHeader": null,
    "entityType": "container",
    "entityUrn": "urn:li:container:0b9f1f731ecf6743be6207fec3dc9cba",
    "entityKeyAspect": null,
    "changeType": "UPSERT",
    "aspectName": "containerProperties",
    "aspect": {
      "value": "{\"customProperties\": {\"platform\": \"glue\", \"instance\": \"PROD\", \"database\": \"flights-database\"}, \"name\": \"flights-database\"}",
      "contentType": "application/json"
    },
    "systemMetadata": null
  },
  {
    "auditHeader": null,
    "entityType": "container",
    "entityUrn": "urn:li:container:0b9f1f731ecf6743be6207fec3dc9cba",
    "entityKeyAspect": null,
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
      "value": "{\"platform\": \"urn:li:dataPlatform:glue\"}",
      "contentType": "application/json"
    },
    "systemMetadata": null
  },
```

In the first one it calls my attention `instance: PROD` because we usually refer that as factory, otherwise it may be confused with the platform instance. But my main concern is: I should probably add the `plaltformInstance` to that `customProperties` or at least include it in the `name`, WDYT?
I understand this issue should be fixed here https://github.com/sgomezvillamor/datahub/blob/master/metadata-ingestion/src/datahub/emitter/mcp_builder.py#L93 However not sure where that `dataclasses` came from, can you help me to understand that?

In the second, it clearly states the aspect is `dataPlatformInstance` however it just mentions `glue`. I need to fix to the platform instance itself.
In this case, the issue is here https://github.com/sgomezvillamor/datahub/blob/master/metadata-ingestion/src/datahub/emitter/mcp_builder.py#L106 Again I need some help to connect the dots here 😅

**Comment by sgomezvillamor** [comment]
Actually I have found some other examples were IMO instance is badly set as the env 🤔 https://github.com/sgomezvillamor/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/sql/athena.py#L107-L109

File: metadata-ingestion/src/datahub/ingestion/source/aws/glue.py
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/aws/glue.py:576]: @rslanka I couldn't make this work when null, the avro converter complains about a None field in the datum not matching the schema.
I'm not an expert but think this is related to the fact that there is an optional field in an middle class in the inheritance https://github.com/sgomezvillamor/datahub/blob/master/metadata-ingestion/src/datahub/emitter/mcp_builder.py#L36

**Comment by sgomezvillamor** [comment]
Solving this

```
  {
    "auditHeader": null,
    "entityType": "container",
    "entityUrn": "urn:li:container:0b9f1f731ecf6743be6207fec3dc9cba",
    "entityKeyAspect": null,
    "changeType": "UPSERT",
    "aspectName": "dataPlatformInstance",
    "aspect": {
      "value": "{\"platform\": \"urn:li:dataPlatform:glue\"}",
      "contentType": "application/json"
    },
    "systemMetadata": null
  },
```

so it includes the platform instance, requires to change this https://github.com/sgomezvillamor/datahub/blob/585aad1aac09516c9e23ab96bf1a751c0c35cdb8/metadata-ingestion/src/datahub/emitter/mcp_builder.py#L99-L107 so `make_dataplatform_instance_urn` is used instead. Not sure if you want me to fix that in this PR because that will impact many other sources. Maybe it's better to address it in a different PR. WDYT?

**Comment by sgomezvillamor** [comment]
I think this is ready to merge (as soon as all tests pass)

I can address [this](https://github.com/linkedin/datahub/pull/4130#issuecomment-1043297195) and [this](https://github.com/linkedin/datahub/pull/4130#issuecomment-1043256427) in separated PRs

**Comment by sgomezvillamor** [comment]
Well... I couldn't wait 😅  and solved last two issues with https://github.com/linkedin/datahub/pull/4130/commits/26ef3ba1762fc491b015f5ec1c9aca13e0c44c7f and https://github.com/linkedin/datahub/pull/4130/commits/59050d47c7ef84786964da13b0349ba0e006774e

So with this I'm done for today. I would appreciate if you could review @rslanka Thanks!

**Comment by sgomezvillamor** [comment]
@rslanka Just noted your approval got dismissed with latest updates, so if you could review again ☺️

**Comment by sgomezvillamor** [comment]
@shirshanka If you could please have a look at this 🙏

**Review by rslanka — APPROVED** [review]
LGTM!

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py:575]: we'll handle this change in a separate PR -> draft PR is here (https://github.com/linkedin/datahub/pull/4279). 
Once that is merged, you shouldn't need to fix these files.

**Review by shirshanka — CHANGES_REQUESTED** [review]
one small change.. and another "wait until other PR is green-lighted"

**Comment by sgomezvillamor** [comment]
Thanks @shirshanka ! Once PR #4279 is merged I will solve the conflicts and address your comment 👍

**Comment by sgomezvillamor** [comment]
@shirshanka As agreed: https://github.com/datahub-project/datahub/pull/4279 got merged and I updated the PR. So this is ready for review again.
CC'ing @treff7es

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py]: It was merged

**Review by treff7es — APPROVED** [review]

**Review by shirshanka — APPROVED** [review]
Lgtm

**Pull request merged**
